waku/v2/protocol/waku_relay.nim

```diff
@@ -23,79 +27,132 @@ logScope:
     ## connection for a protocol string
     ## e.g. ``/wakusub/0.0.1``, etc...
-    ##
+    debug "Incoming WakuRelay connection", connection=conn, protocol=proto
 
-    debug "Incoming WakuRelay connection"
     try:
       await w.handleConn(conn, proto)
     except CancelledError:
       # This is top-level procedure which will work as separate task, so it
       # do not need to propogate CancelledError.
-      trace "Unexpected cancellation in relay handler", conn
-    except CatchableError as exc:
-      trace "WakuRelay handler leaks an error", exc = exc.msg, conn
+      error "Unexpected cancellation in relay handler", conn=conn, error=getCurrentExceptionMsg()
+    except CatchableError:
+      error "WakuRelay handler leaks an error", conn=conn, error=getCurrentExceptionMsg()
 
   # XXX: Handler hijack GossipSub here?
   w.handler = handler
   w.codec = WakuRelayCodec
 
-method initPubSub*(w: WakuRelay) {.raises: [Defect, InitializationError].} =
-  debug "initWakuRelay"
+method initPubSub(w: WakuRelay) {.raises: [InitializationError].} =
+  ## NOTE: This method overrides GossipSub initPubSub method; it called by the 
+  ##  parent protocol, PubSub.
+  debug "init waku relay"
 
-  # after discussions with @sinkingsugar, this is essentially what is needed for
+  # After discussions with @sinkingsugar: This is essentially what is needed for
   # the libp2p `StrictNoSign` policy
   w.anonymize = true
   w.verifySignature = false
   w.sign = false
 
   procCall GossipSub(w).initPubSub()
 
-  w.init()
+  w.initProtocolHandler()
+
+
+proc new*(T: type WakuRelay, 
+          peerManager: PeerManager, 
+          defaultPubsubTopics: seq[PubsubTopic] = @[],
+          triggerSelf: bool = true): WakuRelayResult[T] =
+  
+  proc msgIdProvider(msg: messages.Message): Result[MessageID, ValidationResult] =
+    let hash = MultiHash.digest("sha2-256", msg.data)
+    if hash.isErr():
+      ok(($msg.data.hash).toBytes())
+    else:
+      ok(hash.value.data.buffer)
+
+  var wr: WakuRelay
+  try:
+    wr = WakuRelay.init(
+      switch = peerManager.switch,
+      msgIdProvider = msgIdProvider,
+      triggerSelf = triggerSelf,
+      sign = false,
+      verifySignature = false,
+      maxMessageSize = MaxWakuMessageSize
+    )
+  except InitializationError:
+    return err("initialization error: " & getCurrentExceptionMsg())
+
+  wr.peerManager = peerManager
+  wr.defaultPubsubTopics = defaultPubsubTopics
+
+  ok(wr)
+
+
+method start*(w: WakuRelay) {.async.} =
+  debug "start"
+  await procCall GossipSub(w).start()
+
+method stop*(w: WakuRelay) {.async.} =
+  debug "stop"
+  await procCall GossipSub(w).stop()
 
-method subscribe*(w: WakuRelay,
-                  pubSubTopic: PubsubTopic,
-                  handler: TopicHandler) =
-  debug "subscribe", pubSubTopic=pubSubTopic
 
-  procCall GossipSub(w).subscribe(pubSubTopic, handler)
+method subscribe*(w: WakuRelay, pubsubTopic: PubsubTopic, handler: SubsciptionHandler|PubsubRawHandler) =
```

Comment:
> This results in the baseless methods deprecated... compiler warning.

>I'm not sure, though, how well the dynamic binding will work for the `unsubscribe` and `unsubscribeAll` methods if converted to proc which essentially matches a `proc` in the parent module.

Yes, this was one of the things I left out because I didn't know the impact of those changes.

> I think, however, they can possibly be removed as they don't add more value than some logging before just calling the parent proc?

Not sure about this. I rather leave it for a second round :)